src/index.ts

```diff
@@ -154,50 +146,65 @@ export default class QuickbooksConnector extends BaseHttpConnector<
     return false
   }
 
-  private async refreshTokenIfNeeded() {
-    const dbToken = await this.getQBToken()
-    if (!dbToken) {
-      console.error(
-        `refreshTokenIfNeeded, stored token not found for ${this.configOptions.realmId}`,
-      )
-      return
-    }
-    const expiry = Number(dbToken.token.createdAt) + Number(dbToken.token.expires_in) * 1000
-    const needRefresh = expiry - 120 * 1000 < Date.now() // 2 minutes before token is expired
-
-    if (needRefresh) {
-      console.log('Before refreshing')
-      try {
-        this.oauthClient.setToken(dbToken.token)
-        const authResponse = await this.oauthClient.refresh()
-        await this.storeTokenAndSetClient(authResponse)
-        console.log('Token is refreshed')
-      } catch (e: any) {
-        console.error('Refresh Token Error:', e)
-        console.error(e.intuit_tid)
+  private async refreshTokenWhenNeeded() {
+      const dbToken = await this.getQBToken()
+      if(!dbToken) {
+        this.refreshIsRunning = false
+        return
       }
+
+      const expiry = this.timeToRefreshToken(dbToken)
+      const self = this
+  
+      this._timeout = setTimeout(async function refresh(){
+        await self.refreshToken(dbToken.token)
+        self.refreshTokenWhenNeeded()
+      }, expiry)
+  }
+
+  /**
+   * expiry = createdAt + expires_in - now() - 2 minutes
+   * if expiry is negative - return 0 for imidiate refresh
+   */
+  private timeToRefreshToken(dbToken: any) {
+    let expiry = Number(dbToken.token.createdAt) + Number(dbToken.token.expires_in) * 1000
```

Comment:
Should probably try/catch this and have a default, short time to refresh (in case createdAt or expires_in are not set, or the API changes)

Reply:
Should also have a return type for the function.
Also the last 2 lines can be merged to
`return expiry > 0? expiry : 0`